Userland/Applications/FileManager/PropertiesWindow.cpp

```diff
@@ -269,3 +280,73 @@ ErrorOr<NonnullRefPtr<GUI::Button>> PropertiesWindow::make_button(DeprecatedStri
     button->set_fixed_size(70, 22);
     return button;
 }
+
+void PropertiesWindow::close()
+{
+    GUI::Window::close();
+    if (m_directory_statistics_calculator)
+        m_directory_statistics_calculator->stop();
+}
+
+PropertiesWindow::DirectoryStatisticsCalculator::DirectoryStatisticsCalculator(DeprecatedString path)
+{
+    m_work_queue.enqueue(path);
+}
+
+void PropertiesWindow::DirectoryStatisticsCalculator::start()
+{
+    using namespace AK::TimeLiterals;
+    VERIFY(!m_background_action);
+
+    m_background_action = Threading::BackgroundAction<int>::construct(
+        [this, strong_this = NonnullRefPtr(*this)](auto& task) {
+            auto timer = Core::ElapsedTimer();
+            while (!m_work_queue.is_empty()) {
+                auto base_directory = m_work_queue.dequeue();
+                Core::DirIterator di(base_directory, Core::DirIterator::SkipParentAndBaseDir);
+                while (di.has_next()) {
+                    if (task.is_cancelled())
+                        return ECANCELED;
+
+                    auto path = di.next_path();
+                    struct stat st = {};
+                    if (fstatat(di.fd(), path.characters(), &st, AT_SYMLINK_NOFOLLOW) < 0) {
+                        perror("fstatat");
+                        continue;
+                    }
+
+                    if (S_ISDIR(st.st_mode)) {
+                        auto full_path = LexicalPath::join("/"sv, base_directory, path).string();
+                        m_directory_count++;
+                        m_work_queue.enqueue(full_path);
+                    } else if (S_ISREG(st.st_mode) || S_ISLNK(st.st_mode)) {
+                        m_file_count++;
+                        m_total_size_in_bytes += st.st_size;
+                    }
+
+                    // Show the first update, then show any subsequent updates every 100ms.
+                    if (!task.is_cancelled() && on_update && (!timer.is_valid() || timer.elapsed_time() > 100_ms)) {
+                        timer.start();
+                        on_update(m_total_size_in_bytes, m_file_count, m_directory_count);
+                    }
+                }
+            }
+            return ESUCCESS;
+        },
+        [this](auto result) -> ErrorOr<void> {
+            if (result != ESUCCESS)
+                return {};
+
+            if (on_update && result == 0)
+                on_update(m_total_size_in_bytes, m_file_count, m_directory_count);
+
+            stop();
```

Comment:
is it correct to cancel the job from the success callback? that seems odd to me.

Reply:
No idea why I did this, it's completely unnecessary. I've removed it.